src/irmin-git/irmin_git.ml

```diff
@@ -154,7 +156,43 @@ module Make_private
   end
   module XContents = struct
     include AO (GitContents)
-    module Val = C
+    module Val = struct
+      include C
+
+      let encode_bin buf off (t:t) =
+        Log.debug (fun l -> l "Content.encode_bin");
+        let blob = G.Value.Blob.of_string (Irmin.Type.encode_bin C.t t) in
+        match Raw.to_raw (G.Value.blob blob) with
+        | Error _ -> assert false
+        | Ok s    ->
+          let len = String.length s in
+          Bytes.blit_string s 0 buf off len;
+          off + len
+
+      let decode_bin buf off =
+        Log.debug (fun l -> l "Content.decode_bin");
+        let buf = Cstruct.of_string buf in
+        let buf = Cstruct.shift buf off in
+        let blob t =
+          match Irmin.Type.decode_bin C.t (G.Value.Blob.to_string t) with
+          | Ok t -> t
+          | Error (`Msg e) -> Fmt.failwith "cannot read blob: %s" e
+        in
+        match Raw.of_raw_with_header buf with
+        | Ok (G.Value.Blob t) -> off + Cstruct.len buf, blob t
+        | Ok _    -> failwith "wrong object kind"
+        | Error e -> Fmt.invalid_arg "error %a" Raw.DecoderRaw.pp_error e
+
+      let size_of t =
+        Log.debug (fun l -> l "Content.size_of");
```

Comment:
This is clearly not optimal, but it is hard (impossible?) to know the size after deflate without actually deflating things. So the `bin` API in `Irmin.Type` should probably be updated to allow non-blocking binary serialisation instead of having to require the existence of `size_of`. This could be fixed later, and/or the API could be changed a bit (will try something).